src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -877,6 +917,40 @@ void G1CollectedHeap::resize_heap_if_necessary() {
   }
 }
 
+void G1CollectedHeap::resize_heap_if_necessary_ahs(bool record_expand_time) {
+  assert(SoftMaxHeapSize > 0,
```

Comment:
I'm having trouble understanding the role of SoftMaxHeapSize. What problem is it trying to solve?